[WebAssembly SIMD] Emulate 8-bit shift instructions and i8x16.popcnt on Intel

ddegazio opened the pull request:
#### 7d8a35c36444a89da6244c8122e7bd12780ab3f6
<pre>
[WebAssembly SIMD] Emulate 8-bit shift instructions and i8x16.popcnt on Intel
<a href="https://bugs.webkit.org/show_bug.cgi?id=248995">https://bugs.webkit.org/show_bug.cgi?id=248995</a>
rdar://103159176

Reviewed by Yusuke Suzuki.

Adds implementations for 8-bit shift and popcount SIMD instructions on Intel,
along with i64x2 arithmetic right shift. With this patch, our support for
WebAssembly SIMD on Intel is feature-complete for BBQ Air, and we pass all
spec tests for the Air backend.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::compareFloatingPointVector):
(JSC::MacroAssemblerX86_64::vectorAndnot):
(JSC::MacroAssemblerX86_64::vectorTruncSatSignedFloat64):
(JSC::MacroAssemblerX86_64::vectorUshl8):
(JSC::MacroAssemblerX86_64::vectorUshr8):
(JSC::MacroAssemblerX86_64::vectorSshr8):
(JSC::MacroAssemblerX86_64::vectorPopcnt): Deleted.
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vpunpcklbw_rrr):
(JSC::X86Assembler::vpunpckhbw_rrr):
(JSC::X86Assembler::vcmpps_rrr):
(JSC::X86Assembler::vcmppd_rrr):
(JSC::X86Assembler::vpsllw_i8rr):
(JSC::X86Assembler::vpslld_i8rr):
* Source/JavaScriptCore/b3/air/AirLowerMacros.cpp:
(JSC::B3::Air::lowerMacros):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::addSIMDV_V):
(JSC::Wasm::AirIRGenerator64::addSIMDRelOp):
(JSC::Wasm::AirIRGenerator64::addSIMDShift):

Canonical link: <a href="https://commits.webkit.org/258089@main">https://commits.webkit.org/258089@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12c9a9bad5ac6a946b63e3d5d68f605e1bc524b3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100888 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10041 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33937 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110195 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/170466 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10972 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/910 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93303 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108037 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106671 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8299 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34914 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22954 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77889 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/91373 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3729 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/87463 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/1258 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3747 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29241 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9860 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43970 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/90351 "Built successfully") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5545 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5525 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20226 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3375,6 +3369,75 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+        m_assembler.vpsllw_rrr(shift, tmp2, tmp2);
+
+        // Mask away higher bits of left-shifted results.
+        m_assembler.vpcmpeqw_rrr(dest, dest, dest);
```
You are early-def'ing dest here, but the air opcode only claims to def it

ddegazio replied:
Can scratches interfere with defs? My assumption was that scratch registers would be allocated separately from the other parameters, in which case this should still be fine (since we only write `dest` after we finish using `input` and `shift`). Good catch, I'll check what we do in the register allocator, I guess worst comes to worst we can make `dest` a use/def.

ddegazio replied:
It seems to be the case that scratches are always allocated separately from defs within the same instruction, so this code should work. But, in trying to avoid writing to dest, I ended up finding a different instruction sequence that is probably a bit faster and still passes the tests, so I'll be amending this PR to use that.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
```diff
@@ -1548,9 +1575,39 @@ auto AirIRGenerator64::addSIMDShift(SIMDLaneOperation op, SIMDInfo info, Express
+            // x86 has no SIMD 64-bit signed right shift instruction, so we scalarize it here.
+
+            Tmp shiftRCX = Tmp(X86Registers::ecx);
+            Tmp lower = newTmp(B3::GP), upper = newTmp(B3::GP);
```
nit: separate lines

ddegazio replied:
Split them up in the latest commit.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
```diff
@@ -274,13 +274,40 @@ class AirIRGenerator64 : public AirIRGeneratorBase<AirIRGenerator64, TypedTmp> {
+                // x86_64 does not natively support vector lanewise popcount, so we emulate it using multiple
+                // masks.
+
+                v128_t bottomNybbleConst, popcntConst;
```
nit

ddegazio replied:
Split them to separate lines as well in the latest commit.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2901,19 +2901,19 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
         ASSERT(scalarTypeIsFloatingPoint(simdInfo.lane));
         if (simdInfo.lane == SIMDLane::f32x4)
-            m_assembler.vminps_rrr(right, left, dest);
+            m_assembler.vminps_rrr(left, right, dest);
         else
-            m_assembler.vminpd_rrr(right, left, dest);
+            m_assembler.vminpd_rrr(left, right, dest);
```
Does this mean that x64 and ARM64 have different semantics on Pmin?
If so, let's add comment describing why this order matters.

ddegazio replied:
Added a comment above in the latest push.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2901,19 +2901,19 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
-            m_assembler.vminpd_rrr(right, left, dest);
+            m_assembler.vminpd_rrr(left, right, dest);
     }
 
     void vectorPmax(SIMDInfo simdInfo, FPRegisterID left, FPRegisterID right, FPRegisterID dest)
     {
         RELEASE_ASSERT(supportsAVX());
         ASSERT(scalarTypeIsFloatingPoint(simdInfo.lane));
         if (simdInfo.lane == SIMDLane::f32x4)
-            m_assembler.vmaxps_rrr(right, left, dest);
+            m_assembler.vmaxps_rrr(left, right, dest);
         else
-            m_assembler.vmaxpd_rrr(right, left, dest);
+            m_assembler.vmaxpd_rrr(left, right, dest);
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2932,7 +2932,7 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
         RELEASE_ASSERT(supportsAVX());
         RELEASE_ASSERT(simdInfo.lane == SIMDLane::v128);
-        m_assembler.vandnps_rrr(right, left, dest);
+        m_assembler.vandnps_rrr(left, right, dest); // left and right are intentionally reversed to achieve the desired semantics
```
Can you describe what the semantics difference is in a comment?

ddegazio replied:
Added a comment above in the latest push.

Constellation reviewed (approved):
r=me with comments.

ddegazio committed:
[WebAssembly SIMD] Emulate 8-bit shift instructions and i8x16.popcnt on Intel
https://bugs.webkit.org/show_bug.cgi?id=248995
rdar://103159176

Reviewed by Yusuke Suzuki.

Adds implementations for 8-bit shift and popcount SIMD instructions on Intel,
along with i64x2 arithmetic right shift. With this patch, our support for
WebAssembly SIMD on Intel is feature-complete for BBQ Air, and we pass all
spec tests for the Air backend.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::compareFloatingPointVector):
(JSC::MacroAssemblerX86_64::vectorAndnot):
(JSC::MacroAssemblerX86_64::vectorTruncSatSignedFloat64):
(JSC::MacroAssemblerX86_64::vectorUshl8):
(JSC::MacroAssemblerX86_64::vectorUshr8):
(JSC::MacroAssemblerX86_64::vectorSshr8):
(JSC::MacroAssemblerX86_64::vectorPopcnt): Deleted.
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vpunpcklbw_rrr):
(JSC::X86Assembler::vpunpckhbw_rrr):
(JSC::X86Assembler::vcmpps_rrr):
(JSC::X86Assembler::vcmppd_rrr):
(JSC::X86Assembler::vpsllw_i8rr):
(JSC::X86Assembler::vpslld_i8rr):
* Source/JavaScriptCore/b3/air/AirLowerMacros.cpp:
(JSC::B3::Air::lowerMacros):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::addSIMDV_V):
(JSC::Wasm::AirIRGenerator64::addSIMDRelOp):
(JSC::Wasm::AirIRGenerator64::addSIMDShift):

Canonical link: https://commits.webkit.org/258089@main
